packages/pds/src/pipethrough.ts

```diff
@@ -179,32 +334,111 @@ export const parseProxyHeader = async (
   return { did, serviceUrl }
 }
 
-// Sending request
+// Request parsing/forwarding
 // -------------------
 
-export const makeRequest = async (
-  url: URL,
-  reqInit: RequestInit,
-): Promise<Response> => {
-  let res: Response
+type Accept = [name: string, flags: { q: string }]
+
+function negotiateAccept(
+  acceptHeader: undefined | string | string[],
+  supported: readonly Accept[],
+): readonly Accept[] {
+  // Optimization: if no accept-encoding header is present, skip negotiation
+  if (!acceptHeader?.length) {
+    return supported
+  }
+
+  const acceptNames = extractAcceptedNames(acceptHeader)
+  const common = acceptNames.includes('*')
+    ? supported
+    : supported.filter(nameIncludedIn, acceptNames)
+
+  // There must be at least one common encoding with a non-zero q value
+  if (!common.some(isNotRejected)) {
+    throw new XRPCServerError(
+      ResponseType.NotAcceptable,
+      'NotAcceptable',
+      'this service does not support any of the requested encodings',
+    )
+  }
+
+  return common
+}
+
+function formatAccepted(accept: readonly Accept[]): string {
+  return accept.map(formatEncodingDev).join(', ')
+}
+
+function formatEncodingDev([enc, { q }]: Accept): string {
+  return q != null ? `${enc};q=${q}` : enc
+}
+
+function nameIncludedIn(this: readonly string[], accept: Accept): boolean {
```

Comment:
is there a big performance gain from the `this`/`thisArg` syntax? I find it harder to reason about personally

Reply:
Well, creating an inline function is one additional object to initialize + garbage collect on every incoming HTTP call. JS is arguably good at creating inline function but I really tried to get the best perfs where I could.

I don't have a bench to back me though...